waspc/src/Wasp/AppSpec/Valid.hs

```diff
@@ -49,6 +56,55 @@ validateExactlyOneAppExists spec =
         GenericValidationError $
           "You have more than one 'app' declaration in your Wasp app. You have " ++ show (length apps) ++ "."
 
+validateWasp :: AppSpec -> [ValidationError]
+validateWasp spec = validateWaspVersion specWaspVersionStr
+  where
+    specWaspVersionStr = Wasp.version $ App.wasp (snd $ getApp spec)
+
+validateWaspVersion :: String -> [ValidationError]
+validateWaspVersion specWaspVersion = case versionMatches of
+  [] -> [GenericValidationError "Wasp version should be in the format ^0.major.minor.patch"]
+  versionDigitsStr -> validateVersionDigits versionDigitsStr
+  where
+    (_ :: String, _ :: String, _ :: String, versionMatches :: [String]) = specWaspVersion =~ versionRegex
+    versionRegex = "\\^0\\.([0-9]+).([0-9]+).([0-9]+)$" :: String
+
+    validateVersionDigits :: [String] -> [ValidationError]
+    validateVersionDigits versionDigits = validateCompatibility expectedSemVer currentWaspSemVer
+      where
+        expectedSemVer = wordsToSemVer $ map read versionDigits
+
+        wordsToSemVer :: [Word] -> SV.Version
+        wordsToSemVer (major : minor : patch : _) = SV.Version (wordToNatural major) (wordToNatural minor) (wordToNatural patch)
+        wordsToSemVer _ =
+          error $
+            "Expected version digits to match [major, minor, patch]. "
+            ++ "This should never happen as the version was already validated with regex."
+
+        validateCompatibility :: SV.Version -> SV.Version -> [ValidationError]
+        validateCompatibility compatibleVersion actualVersion =
+          if isCurrentWaspVersionCompatibleWithSpecVersion
+            then []
+            else
+              [ GenericValidationError $
+                  incompatibleVersionMessage actualVersion compatibleVersion
+              ]
+          where
+            isCurrentWaspVersionCompatibleWithSpecVersion =
+              SV.isVersionInRange actualVersion (SV.Range [SV.backwardsCompatibleWith compatibleVersion])
+
+        incompatibleVersionMessage :: SV.Version -> SV.Version -> String
+        incompatibleVersionMessage actualVersion expectedVersion =
+          unwords
+            [ "Your Wasp version does not match the app's requirements.",
+              "You are running Wasp 0." ++ show actualVersion ++ ".",
+              "This app requires Wasp ^0." ++ show expectedVersion ++ "."
+            ]
+
+        currentWaspSemVer = SV.Version (toEnum currentMajor) (toEnum currentMinor) (toEnum currentPatch)
+          where
+            DV.Version (_ : currentMajor : currentMinor : currentPatch : _) _ = Paths_waspc.version
```

Comment:
Great work here, especially taking into account you are relatively new to Haskell!!!

You did a pretty good job here, but I had a feeling couple things might be simplified. One thing that I notice first is three levels of `where` -> that is valid, but is a bit unusual and it does make reading the function a bit harder. Yeah, I had a bit hard time reading the function, although it wasn't badly written, it just felt a bit too verbose actually.

I also noticed you used `Word` -> You can instead directly read those version digits into Natural, `read` adapts to read whatever you need.

Finally, I wasn't sure what happens when you call `toEnum` to convert Int to Natural, so I implemented function for it instead.

Then I reorganized it all a bit also, and I made it clear we are dealing with pre-release versions.

Here is the version I ended up with:
```hs
validateWaspVersion :: String -> [ValidationError]
validateWaspVersion specWaspVersionStr = eitherUnitToErrorList $ do
  specWaspPreReleaseVersion <- parseWaspPreReleaseVersion specWaspVersionStr
  unless
    ( SV.isVersionInRange currentWaspPreReleaseVersion $
        SV.Range [SV.backwardsCompatibleWith specWaspPreReleaseVersion]
    )
    $ Left $ incompatiblePreReleaseVersionError currentWaspPreReleaseVersion specWaspPreReleaseVersion
  where
    -- Given a wasp pre-release version (meaning it is not production ready yet,
    -- so it starts with 0) of format '^0.major.minor.patch`, it returns semver
    -- version major.minor.patch .
    parseWaspPreReleaseVersion :: String -> Either ValidationError SV.Version
    parseWaspPreReleaseVersion waspVersionStr = do
      let (_ :: String, _ :: String, _ :: String, waspVersionDigits :: [String]) =
            waspVersionStr =~ ("\\^0\\.([0-9]+).([0-9]+).([0-9]+)$" :: String)
      case mapM readMaybe waspVersionDigits of
        Just [major, minor, patch] -> Right $ SV.Version major minor patch
        __ -> Left $ GenericValidationError "Wasp version should be in the format ^0.major.minor.patch"

    incompatiblePreReleaseVersionError :: SV.Version -> SV.Version -> ValidationError
    incompatiblePreReleaseVersionError actualVersion expectedVersion =
      GenericValidationError $
        unwords
          [ "Your Wasp version does not match the app's requirements.",
            "You are running Wasp 0." ++ show actualVersion ++ ".",
            "This app requires Wasp ^0." ++ show expectedVersion ++ "."
          ]

    currentWaspPreReleaseVersion :: SV.Version
    currentWaspPreReleaseVersion = SV.Version (intToNat major) (intToNat minor) (intToNat patch)
      where
        DV.Version [0, major, minor, patch] _ = Paths_waspc.version
        intToNat = naturalFromInteger . toInteger

    eitherUnitToErrorList :: Either e () -> [e]
    eitherUnitToErrorList (Left e) = [e]
    eitherUnitToErrorList (Right ()) = []

```

I thought this reads a bit easier, what do you think?

Regardless of all of this, I had another thought -> a lot of complexity here is coming from the fact that we have this special case of `0.x.y.z` version. This is not standard Semver, standard Semver is `0.x.y` where they treat `x` as major and `y` as minor. This is my fault -> I for some reason introduced this non-standard way for versioning some time ago.
Now that I think about it, I don't think we gained much by doing `0.x.y.z`, and we could probably benefit from switching to `0.x.y` which is standard! What do you think?
So what I would propose, is that we actually refactor now this PR assuming that we use `0.x.y` for versioning, and that the version under which this code will be next released is `0.7.0`. While doing this reactoring you can look at your old code, my code that I proposed above, and try to get the best out of that all.
That should simplify this piece of code, and also make our lives easier in the future!

Btw we are actually handling correctly case in SemVer when first digit is 0, in our SemanticVersion.hs, so that is already taken care of.

Reply:
Thank you so much! Your refactor is so much better than my implementation. I learned a lot just by reading it :smile: 

I think changing the versioning scheme to regular SemVer would be really nice! It would clear up any confusion that may arise about Wasp's version (actually, I was confused about this when I started working on the issue).

What should we do about the tests, though? They will fail until the project's version is changed to `0.7.0`. Unless we change the version in the `cabal` file in this same PR.

> standard Semver is 0.x.y where they treat x as major and y as minor

I think SemVer specifies `0.y.z` to be `major: 0`, `minor: y`, `patch: z`. https://semver.org/#spec-item-4

Let me know what you think :smile:

Reply:
I am glad you found it useful! I think your implementation was already pretty good though, I mostly did some reorganizing + a bit of monads (Either).

Regarding the version: let's do the following then, let's set project's version from `0.6.0.0` to `0.6.0`, in the project (so in .cabal file, and maybe some other files).
Then once we do new release, we will increase it to 0.7.0 -> but that is not your concern right now.

SemVer and 0.x.y -> you are right, those are still called minor and patch, which is why I said "they treat x as major and y as minor", not that they are actually major and minor. You can check this out here: https://github.com/npm/node-semver#caret-ranges-123-025-004  -> x is considered to be introducing breaking changes, while y is considered to be doing additions.

All together, sounds good to me and let's go with `0.6.0`!

Reply:
I just applied the corresponding changes. Thank you for your help :D